util/encoding.go

```diff
@@ -0,0 +1,243 @@
+package util
+
+import (
+	"encoding/base64"
+	"errors"
+	"unsafe"
+)
+
+var (
+	fibEncodeNumOutOfRangeErr = errors.New("the number to be encoded is out of range")
+	fibEncodeInvalidRange     = errors.New("the range is invalid")
+)
+
+func getByteSlice() []byte {
+	// Most string to be encoded are less than 8 bytes.
+	return make([]byte, 0, 8)
+}
+
+func NewBitStreamForWrite() *BitStream {
+	return NewBitStream(getByteSlice())
+}
+
+// enlarge the underlying byte slice.
+// In short, if the final size is less than the expThreshold, double the size of the underlying byte array,
+// otherwise, enlarge the size of the array by a specified step.
+func (bs *BitStream) enlarge(n uint16) {
+	final := int(bs.p+n+7) / 8
```

Comment:
This is calculating the final position (or size) from where the bitstream pointer currently is. It would be sketchy to write to the middle of a bit stream, but should probably make this clearer in the comment so someone doesn't assume it can divine where the end of the bit stream is.